[Feature] Add ConceptGraph

neenjaw opened the pull request:
This is a translation of [neenjaw/ruby-exercism-graph](https://github.com/neenjaw/ruby-exercism-graph) according to neenjaw/ruby-exercism-graph#1.

Purpose: The Concept Graph of the v3-dashboard created an important visualization for the progression of the track as it is to be presented to students.  Non-final versions of this are being worked on for the student-facing website.  This is a small class library to be able to transform a list of concept-exercise hashes (not ActiveRecords) into a directed graph, and also to compute a presentation layout.

Four files are being added here (2 source with 2 test):
- `app/models/concept_graph.rb`
  - The object creates the graph from the list of concept-exercise hashes
- `lib/graph_utils.rb`
  - provides module methods for working with graph data structures
- `test/app/models/concept_graph_test.rb`
- `test/lib/graph_utils.rb`

Things to work on:
- how the concept graph obtains its data & when, how it is triggered
- test coverage
  - Being newer to rails, I am unsure how to test this feature in isolation from the other tests (I don't have mariadb set up locally and my attempts to run tests have failed because I think it can't connect to a test database)
  - How are test fixtures defined in rails tests?
    - Just so I can test a known list of exercises and determine the expected results

neenjaw committed:
Adding refactored files

iHiD committed:
Get graph utils tests working correctly.

Anything in the lib/ directly in Rails needs requiring.  Everything in the models (or other directories) gets automatically loaded.

iHiD committed:
Get other tests working

iHiD committed:
Convert to a command

iHiD committed:
Remove language key

iHiD committed:
Refactor graph code

neenjaw committed:
Change `nodes_key_by_concept`

Change to use `flat_map {_}.to_h` rather than `each_with_object({})`

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -0,0 +1,121 @@
+      Node = Struct.new(:index, :slug, :uuid, :concepts, :prerequisites, keyword_init: true)
+
+      # Add public readers
+      attr_reader :nodes, :edges
```
Switch this to be private now.

neenjaw replied:
Done, see next commit

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -0,0 +1,121 @@
+        nodes_keyed_by_concept[concept]
+      end
+
+      def node_for_index(index)
```
We don't publically use this, so we could make it private or just use `nodes_keyed_by_index` directly internally.

neenjaw replied:
`#node_for_index(idx)` is used in `DetermineConceptsLayout#call` on line 23, did you mean `#node_for_concept(concept)`?

iHiD replied:
Yes. Sorry 🙂

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -0,0 +1,121 @@
+            prerequisites: exercise['prerequisites']
+          )
+        end.freeze
+      end
```
Just a future note: We won't actually need this method when we pass real data in, as the exercises will already be things that have all of these methods defined on them. So the constructor will just be able to have `@nodes = exercises`

neenjaw replied:
I'm uncertain which segment of the code you're referring to here by the snippet shown, but I think what you are saying is that the `Node` class may be redundant at some point in the future.  One limitation would be that an arbitrary index is assigned to the node in this implementation to facilitate graph manipulation since the graph functions use the index of the array to represent the node in the graph.

iHiD replied:
OK, cool. Yeah - that's good to know 🙂

iHiD replied:
(For clarity, I meant the whole of this method above).

iHiD commented on test/commands/track/determine_concepts_layout_test.rb:
```diff
@@ -0,0 +1,45 @@
+            'prerequisites' => []
+          }
+        ]
+      }.freeze
```
I wouldn't bother changing this, as we'll run this of factories before we merge anyway, but this would be more Rubyish to use symbols for the keys, rather than strings and fat arrows. e.g.:

```
            slug: 'basics',
            uuid: 'c29c6092-9d44-4f21-8138-b873384fd90b',
            concepts: %w[basics],
            prerequisites: []
          }
```

iHiD commented on test/lib/graph_utils_test.rb:
```diff
@@ -0,0 +1,30 @@
+  def test_detect_cycles
+    assert GraphUtils.contains_cycle?([[1], [0]])
+    assert GraphUtils.contains_cycle?([[1], [2], [0]])
+    assert GraphUtils.contains_cycle?([[1], [2], [3], [1]])
```
Can we have a `refute` or two in here too pls. Atm the test would pass if it `contains_cycle?` just returned `true`.

neenjaw replied:
done

neenjaw committed:
proper exception added, graph class vars private

neenjaw committed:
chain of variables to `then`

iHiD commented on config/initializers/exceptions.rb:
```diff
@@ -6,3 +6,4 @@ class SolutionLockedByAnotherMentorError < RuntimeError; end
 class TooManyIterationsError < RuntimeError; end
 class TrackSearchStatusWithoutUserError < RuntimeError; end
 class TrackSearchInvalidStatusError < RuntimeError; end
+class TrackHasCyclicPrerequisiteError < RuntimeError; end
```
Good name. Could you put this in alphabetical order pls :)

neenjaw committed:
sort Errors alphabetically

neenjaw committed:
rejigger to also get connections on the graph

neenjaw committed:
Apply suggestions from code review from @iHiD

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>

neenjaw committed:
Changes/fixes
- add :level to struct
- use flat_map instead of reduce
- use prereq's :level rather than make previous level lookup

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -0,0 +1,117 @@
+  class DetermineConceptsLayout
+    include Mandate
+
+    def initialize(track_config)
```
So, presuming we pass a `track` into this, we have:

```ruby
track.concept_exercises #=> [Track::ConceptExercise]

# Then for each concept exercise
concept_exercise.taught_concepts #=> [Track::Concept]
concept_exercise.prerequisites #=> [Track::Concept]
```

You can get these set up in the tests using something like:

```ruby
track = create track
strings = create :track_concept, slug: "strings", track: track
booleans = create :track_concept, slug: "booleans", track: track

e1 = create :concept_exercise
e1.taught_concepts << strings

e2 = create :concept_exercise
e2.taught_concepts << booleans

e3 = create :concept_exercise
e3.prerequisites << strings
e3.prerequisites << booleans
```

This currently relies on #152 being merged.

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -0,0 +1,142 @@
+          Node.new(
+            index: idx,
+            slug: exercise['slug'],
+            uuid: exercise['uuid'],
```
Two thoughts:
1. I think it would be worth passing in the Concept object (which you'll get from the code above: https://github.com/exercism/v3-website/pull/112/files#r499166532). We can then use that to generate URLs, get their names, etc.
2. We don't use this uuid field anywhere as far as I can see.

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -0,0 +1,142 @@
+        level.flat_map do |node|
+          node.prerequisites.
+            map { |prerequisite| graph.node_for_concept(prerequisite) }.
+            reject(&:nil?).
```
```suggestion
            compact.
```

iHiD replied:
(This is purely aesthetic)

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -0,0 +1,142 @@
+      raise TrackHasCyclicPrerequisiteError if graph.has_cycle?
+
+      Array.new(graph.levels.max + 1) { [] }.tap do |level|
+        graph.levels.each_with_index do |level_idx, node_idx|
```
```suggestion
        graph.levels.each.with_index do |level_idx, node_idx|
```

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -0,0 +1,142 @@
+
+    memoize
+    def connections
+      levels.drop(1).with_index.flat_map do |(level, level_idx)|
```
```suggestion
      levels.drop(1).each.with_index.flat_map do |level, level_idx|
```

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -0,0 +1,142 @@
+            map { |prerequisite| graph.node_for_concept(prerequisite) }.
+            reject(&:nil?).
+            uniq.
+            select { |prerequisite_node| prerequisite_node.level + 1 == level_idx }.
```
```suggestion
            select { |prerequisite_node| prerequisite_node.level == level_idx }.
```

neenjaw committed:
Apply suggestions from code review from @iHiD to get things working

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>

neenjaw committed:
Graph now accepts ActiveRecord
- `levels` and `connections` now `exercise_*`
- node now does not track uuids
- nodes' `concepts` are now `taught_concepts`

neenjaw committed:
determines concept levels
now is able to determine concept levels from the track/exercise/concept
ActiveRecord model, returning it according to the react component
interface.

Test included for same.

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -7,20 +7,56 @@ class DetermineConceptsLayout
     include Mandate
 
     def initialize(track)
+      @track = track
```
I think currently you don't need to store this as an instance variable.

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -7,20 +7,56 @@ class DetermineConceptsLayout
+
+    memoize
+    def concepts
+      counter = 0 # TODO: remove this if the component doesn't actually need an 'index'
```
Does it? :)

neenjaw replied:
It requires a refactor of the component, maybe in a refinement PR?

neenjaw replied:
Nevermind, I need to add the display name to the component. Will do it in next commit.

iHiD replied:
It's not a blocker at all. I just wondered from the comment what was driving that decision.

neenjaw replied:
I think just a remnant of code from a time when putting together the component. So was writing this around it.  It has now been removed in the component, so removed here.

neenjaw committed:
fix capitalization error
names of vars were changed to satisfy rubocop, must have
forgotten to run tests.

neenjaw committed:
concepts-map update
-unused index removed
-name added

determine_concepts_layout.rb
- unused index generation removed from concepts

iHiD commented on app/commands/track/determine_concepts_layout.rb:
```diff
@@ -26,13 +26,9 @@ def call
 
     memoize
     def concepts
-      counter = 0 # TODO: remove this if the component doesn't actually need an 'index'
```
Nice commit! :)

neenjaw replied:
:smile:

neenjaw committed:
fix js test

neenjaw committed:
test data updated to new form

neenjaw committed:
indexes removed

neenjaw committed:
Component added to view

neenjaw committed:
TLDR: the concept page now shows the concepts map
Several changes made to get this working:
- component made congruent to the ruby code, `layout` -> `levels`
- joined.html.haml template now calls component
- concepts_contoller now calls the `determine_concepts_layout` command

- With database data, an edge is being created as `nil` for the tracks,
  added `#compact` to line 134 of
  "app/commands/track/determine_concepts_layout.rb"
  - need to figure out why

neenjaw committed:
Bugfix: compact nil edges when they are created
